drivers/modem/ublox-sara-n310.c

```diff
@@ -0,0 +1,526 @@
+				    sizeof(minfo.mdm_revision) - 1,
+				    data->rx_buf, 0, len);
+	minfo.mdm_revision[out_len] = '\0';
+
+	return 0;
+}
+
+MODEM_CMD_DEFINE(on_cmd_atcmdinfo_imei)
+{
+	size_t out_len;
+
+	out_len = net_buf_linearize(minfo.mdm_imei, sizeof(minfo.mdm_imei) - 1,
+				    data->rx_buf, 0, len);
+	minfo.mdm_imei[out_len] = '\0';
+
+	return 0;
+}
+
+MODEM_CMD_DEFINE(on_cmd_cpgaddr)
+{
+	size_t out_len;
+	char buf[sizeof(minfo.mdm_ip)];
+
+	out_len = net_buf_linearize(buf, sizeof(buf) - 1, data->rx_buf, 0, len);
+	buf[out_len] = '\0';
+
+	/* extract IP by finding the enclosing quotation marks */
+	char *p = strchr(buf, '\"');
+
+	if (p) {
+		size_t len = strlen(p + 1);
+
+		memmove(buf, p + 1, len + 1);
+
+		p = strchr(buf, '\"');
+		if (p) {
+			len = p - buf;
+
+			memcpy(minfo.mdm_ip, buf, len);
+		}
+	}
+
+	return 0;
+}
+
+/* response command handling */
+static const struct modem_cmd response_cmds[] = {
+	MODEM_CMD("OK", on_cmd_ok, 0U, ""),
+	MODEM_CMD("ERROR", on_cmd_error, 0U, ""),
+	MODEM_CMD("+CCID: ", on_cmd_atcmdinfo_iccid, 0U, ""),
+	MODEM_CMD("+USORF: ", on_cmd_usorf, 1U, ""),
+	MODEM_CMD("+CGPADDR: ", on_cmd_cpgaddr, 1U, ""),
+};
+
+/* unsolicited command handling */
+static const struct modem_cmd unsol_cmds[] = {
+	MODEM_CMD("+UUSORD: ", on_cmd_socknotifydata, 2U, ","),
+	MODEM_CMD("+UUSORF: ", on_cmd_socknotifydata, 2U, ","),
+	MODEM_CMD("+CEREG: ", on_cmd_socknotifycereg, 1U, ""),
+};
+
+/* receive command */
+static void n310_recv(void)
+{
+	/* loop indefinitely in a k_thread */
+	while (1) {
+		/* wait until there is data in RX buffer */
+		k_sem_take(&mdata.mdm_data.rx_sem, K_FOREVER);
+
+		mdata.context.cmd_handler.process(&mdata.context.cmd_handler,
+						  &mdata.context.iface);
+
+		k_yield();
+	}
+}
+
+/* send command */
+void n310_send(uint8_t *payload)
+{
+	if (has_ip()) {
+		int ret = 0;
+
+		LOG_INF("\033[0;37mSending command: %s", payload);
+		ret = modem_cmd_send(&mdata.context.iface,
+				     &mdata.context.cmd_handler, NULL, 0U,
+				     payload, &mdata.sem_response,
+				     K_SECONDS(1));
+
+		if (ret < 0) {
+			LOG_ERR("send function failed");
+		}
+	}
+}
+
+/* pin init */
+static int pin_init(void)
+{
+	/* set and enable pins/pwr, blank for now */
+	return 1;
+}
+
+/* create socket */
+int create_socket(int type, int port)
+{
+	int ret = 0;
+	char buf[sizeof("AT+USOCR=##,#####\r")];
+
+	snprintk(buf, sizeof(buf), "AT+USOCR=%d,%d", type, port);
+
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, buf, &mdata.sem_response, K_SECONDS(4));
+
+	modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler, NULL,
+		       0, "AT+CGPADDR=", &mdata.sem_response, K_SECONDS(2));
+
+	return ret;
+}
+
+/* close socket */
+static int close_socket(void)
+{
+	int ret = 0;
+
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, "AT+USOCL=1", &mdata.sem_response,
+			     K_SECONDS(4));
+
+	return ret;
```

Comment:
I don't know what rules zephyr has about this, but in this case I would remove ret and just do "return modem_cmd_send(etc..)"

Reply:
It is allowed, but I don't see where this would be applicable without missing important steps in this case.